# Data Modeling Seasonality

Now that we've explored what the seasonality dataset looks like, the goal is to be able to make modifications to our average prediction from the initial regression to account for seasonality of pricing. This basically means that we will make +/- modifications to our average prediction based on the the day that we are projecting the data for. We can also do similar things for months of the year as well as holidays. This will hopefully reduce the residuals because the seasonality of the pricing data would cause some correlation among the residuals (based on time of the year) violating a lot of the OLS assumptions. We use averages as a way to explore seasonality. More advanced seasonality measurements could be used if we had more data over several years (where we could build an ARIMA or SARIMA model).

In [37]:
import pandas as pd
import numpy as np
import matplotlib
import seaborn as sb
import matplotlib.pyplot as plt
import matplotlib.cm as cmx
import matplotlib.colors as colors
from sklearn import linear_model
import sklearn.metrics as metrics
from sklearn.grid_search import GridSearchCV
from sklearn import preprocessing
from sklearn.cross_validation import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression as Lin_Reg
from sklearn.linear_model import RidgeCV
from sklearn.linear_model import LassoCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
import pylab
import scipy.stats as stats
%matplotlib inline
import datetime as dt
from datetime import datetime

### Process Overview:
The general idea behind this analysis is as follows: we aggregate prices by weekday for each listing. Then, we normalize each listing's price by the monday price to find an average multiplier for each listing for each day. Then, for each day we average across all listings to get a final average multiplier for each day. Lastly, we compare these predictions to a subset of the listings.

In [21]:
#Importing Datafile
results_nona = pd.read_csv('../datasets/seasonality_tomodel.csv')
results_multiplier = pd.read_csv('../datasets/seasonality_tomodel.csv')
b=['Mon','Tue','Wed','Thu','Fri','Sat','Sun']
for i in b[1:7]:
    results_multiplier[i] = results_multiplier[i]/results_multiplier['Mon']
results_multiplier['Mon']= 1
b=['Mon','Tue','Wed','Thu','Fri','Sat','Sun']
for i in b[1:7]:
    results_multiplier[i] = results_multiplier[i]/results_multiplier['Mon']
results_multiplier['Mon']= 1
results_multiplier.head(5)

Mon       Tue       Wed       Thu       Fri       Sat      Sun  listing_id
0    1  1.000000  1.000000  1.000000  1.000000  1.000000  1.00000   3604481.0
1    1  1.000000  1.000000  1.000000  1.000000  1.000000  1.00000   2949128.0
2    1  0.991826  0.991826  0.999846  1.138965  1.138965  1.00000   4325397.0
3    1  1.000000  1.000000  1.000000  1.000000  1.000000  1.00000   4325398.0
4    1  0.991494  0.994952  1.004395  1.015027  1.011263  0.99895   3426149.0

We see that the dataframe now contains a multiplier for each day of the week for each listing. Now we take an average for each day(averaging across all listings) to see an average multiplier value for each day

In [108]:
multiplier = dict.fromkeys(b)
for index,i in enumerate(multiplier):
    multiplier[i]=results_multiplier.mean()[i]
multiplier

{'Fri': 1.0306309142782333,
 'Mon': 1.0,
 'Sat': 1.0306744793102525,
 'Sun': 1.0009269127770359,
 'Thu': 1.0003671343632679,
 'Tue': 0.99980071473169119,
 'Wed': 0.9998361380714712}

In [115]:
b[1:]

['Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']

The results are very much in line with what we saw earlier in our seasonality-exploration file. Monday and Tuesday see a slight dip in their prices(99.9%) while Friday and Saturday see a sizable increase in prices (103%). These are thus the numbers we will be using to apply seasonality to the averages from our previous models. Here, we apply it to RidgeCV the best one that we found.

## Predicting prices using our seasonality averages

Now, it is important to test the performance of the averages we arrived at

In [31]:
#Reproducing the RidgeCV code from 3A
# helper function for plotting residual plots
def plot_residual(ax1, ax2, ax3, y_pred, y_real, line_label, title):
    ax1.scatter(y_pred, 
                y_real, 
                color='blue',
                alpha=0.6,
                label=line_label)
    ax1.set_xlabel('Predicted Y') 
    ax1.set_ylabel('Real Y')
    ax1.legend(loc='best')
    ax1.set_title(title)

    ax2.scatter(y_pred,
                y_real - y_pred, 
                color='green',
                marker='x',
                alpha=0.6,
                label='Residual')
    ax2.set_xlabel('Predicted Y')
    ax2.set_ylabel('Residual')
    
    ax2.axhline(y=0, color='black', linewidth=2.0, alpha=0.7, label='y=0')

    ax2.legend(loc='best')
    ax2.set_title('Residual Plot')
    
    ax3.hist(y_real - y_pred, bins=30, color='green', alpha=0.7)
    ax3.set_title('Histogram of residual values')
    
    return ax1, ax2, ax3
class model:
    
    def __init__(self, model):
        self.model = model
        self.x_train = None
        self.y_train = None
        self.x_test = None
        self.y_test = None
        self.y_pred_train = None
        self.y_pred_test = None
        self.train_score = None
        self.test_score = None
        self.train_score_log = None
        self.test_score_log = None
    
    def data_split(self, x, y, test_size):
        self.x_train, self.x_test, self.y_train, self.y_test = train_test_split(x, y, test_size=test_size)
    
    def score_reg(self):
        return self.train_score, self.test_score
    
    def score_log(self):
        self.train_score_log = metrics.r2_score(np.exp(self.y_train), np.exp(self.y_pred_train))
        self.test_score_log = metrics.r2_score(np.exp(self.y_test), np.exp(self.y_pred_test))
        return self.train_score_log, self.test_score_log
    
    def data_frame_convert(self):
        df_train = pd.DataFrame({'y_pred': self.y_pred_train, 'y_real': self.y_train})
        df_test = pd.DataFrame({'y_pred_test': self.y_pred_test, 'y_real_test': self.y_test})
        return self.train_score, self.test_score, df_train, df_test

    def data_frame_convert_log(self):
        df_train = pd.DataFrame({'y_pred': np.exp(self.y_pred_train), 'y_real': np.exp(self.y_train)})
        df_test = pd.DataFrame({'y_pred_test': np.exp(self.y_pred_test), 'y_real_test': np.exp(self.y_test)})
        return self.train_score_log, self.test_score_log, df_train, df_test
    
    def fit_model(self, x, y, test_size):
        self.data_split(x, y, test_size)
        self.model = self.model.fit(self.x_train, self.y_train)
        self.train_score = self.model.score(self.x_train, self.y_train)
        self.test_score = self.model.score(self.x_test, self.y_test)
        self.y_pred_train = self.model.predict(self.x_train)
        self.y_pred_test = self.model.predict(self.x_test)
    
def model_iterations(n, x, y, model_arg, log_bool=False):
    training_scores = [None]*n
    testing_scores = [None]*n

    for i in range(n):
        new_model = model(model_arg)
        new_model.fit_model(x, y, 0.3)
        training_scores[i], testing_scores[i] = new_model.score_reg() if not log_bool else new_model.score_log()

    print 'Mean Train Score:', np.mean(training_scores)
    print 'Mean Test Score:', np.mean(testing_scores)
    return new_model

In [32]:
def median_absolute_errors(x, y, log_bool=None):
    reg_params = 10.**np.linspace(-10, 5, 10)
    models = [ RidgeCV(alphas=reg_params, cv=5)]
    model_labels = np.array([ 'RidgeCV'])
    model_errors = np.array([])

    X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.35, random_state=15)

    for model in models:
        model.fit(X_train, y_train)
        if not log_bool:
            model_err = metrics.median_absolute_error((y_test), model.predict(X_test))
            model_errors = np.append(model_errors, model_err)
        else:
            model_err = metrics.median_absolute_error(np.exp(y_test), np.exp(model.predict(X_test)))
            model_errors = np.append(model_errors, model_err)
    
    model_position = np.arange(model_errors.shape[0])
    models_sorted = np.argsort(model_errors)
    for i, model in enumerate(model_labels):
        print 'Model {} Results: {}'.format(model_labels[i], model_errors[i])

    plt.figure(figsize=(10,8))
    plt.bar(model_position, model_errors[models_sorted], align='center')
    plt.xticks(model_position, model_labels[models_sorted])
    plt.xlabel('Estimator')
    plt.ylabel('Median Absolute Error')
    plt.show()

In [33]:
data = pd.read_csv('../datasets/listings_clean.csv')
data.head()
# split into x and y (note that we do not include id and host_id as predictors)
x = data.iloc[:, 2:-2]
y = data.iloc[:, -2]
y_log = data.iloc[:, -1]

In [44]:
data.head(5)

id   host_id  accommodates  bathrooms  bedrooms      beds  \
0  1069266   5867023     -0.520266  -0.331542 -0.407402 -0.493039   
1  2061725   4601412     -0.520266  -0.331542 -0.407402  0.381672   
2    44974    198425     -0.520266  -0.331542 -0.407402 -0.493039   
3  4701675  22590025     -0.520266  -0.331542 -0.407402  0.381672   
4    68914    343302      1.690892  -0.331542  1.266328  1.256383   

   minimum_nights  availability_30  number_of_reviews  host_listing_count  \
0        0.173906         0.390393           2.716107           -0.355961   
1        0.173906        -0.965897           1.295605            0.933455   
2        2.889531        -1.205242           0.822104           -0.355961   
3       -0.601986         1.108429          -0.493176           -0.355961   
4       -0.214040        -0.407424           0.295992            0.073844   

     ...      50-59  60-69  70-79  80-84  85-89  90-94  95-100  No Reviews  \
0    ...          0      0      0      0      1      0       0           0   
1    ...          0      0      0      0      0      0       1           0   
2    ...          0      0      0      0      0      0       1           0   
3    ...          0      0      0      0      0      0       1           0   
4    ...          0      0      0      0      0      0       1           0   

   price  price_log  
0  160.0   5.075174  
1   58.0   4.060443  
2  185.0   5.220356  
3  195.0   5.273000  
4  165.0   5.105945  

[5 rows x 226 columns]

In [73]:
reg_params = 10.**np.linspace(-10, 5, 10)
RidgeCV_model = RidgeCV(alphas=reg_params, fit_intercept=True, cv=5)

In [74]:
RidgeCV_model.fit(x,y_log)

RidgeCV(alphas=array([  1.00000e-10,   4.64159e-09,   2.15443e-07,   1.00000e-05,
         4.64159e-04,   2.15443e-02,   1.00000e+00,   4.64159e+01,
         2.15443e+03,   1.00000e+05]),
    cv=5, fit_intercept=True, gcv_mode=None, normalize=False, scoring=None,
    store_cv_values=False)

In [168]:
sample = results_nona.sample(frac=0.4,axis=0)
len(sample)

1092

In [169]:
# some of the id's in the sample can't be found. So at the end we readjust the sample dataframe too so they have the same entries
sample_variables=data.loc[data['id'].isin(sample['listing_id'])]
sample_variables.head(5)
sample_variables.shape
sample = sample.loc[sample['listing_id'].isin(sample_variables['id'])]

In [170]:
sample_variables.shape

(958, 226)

In [171]:
X_sample = sample_variables.iloc[:, 2:-2]
#note y_test is the un-exponentiated price

In [172]:
new_predictions = sample.copy()
new_predictions.loc[:,0:7]=0

In [173]:
new_predictions['Mon']=np.exp(RidgeCV_model.predict(X_sample))
for i in b[1:]:
    new_predictions[i]=np.exp(RidgeCV_model.predict(X_sample))*multiplier[i]

In [174]:
new_predictions

Mon         Tue         Wed         Thu         Fri         Sat  \
875    38.752293   38.744570   38.745943   38.766520   39.939311   39.941000   
787   117.820157  117.796677  117.800850  117.863412  121.429096  121.434229   
2093   93.586445   93.567795   93.571110   93.620804   96.453083   96.457161   
2721   95.829400   95.810303   95.813698   95.864583   98.764742   98.768917   
2007  246.199927  246.150863  246.159585  246.290316  253.741256  253.751982   
1795  203.023346  202.982886  202.990078  203.097883  209.242137  209.250981   
1030  199.235418  199.195713  199.202771  199.308564  205.338181  205.346861   
1623  134.245922  134.219168  134.223924  134.295208  138.357997  138.363845   
1662  163.660015  163.627400  163.633198  163.720101  168.673071  168.680201   
2089  151.812414  151.782160  151.787538  151.868150  156.462567  156.469181   
459   200.984900  200.944846  200.951966  201.058688  207.141251  207.150007   
696   136.855673  136.828400  136.833248  136.905918  141.047688  141.053650   
2557   83.278746   83.262150   83.265100   83.309321   85.829650   85.833278   
90     62.558593   62.546126   62.548342   62.581561   64.474820   64.477545   
284   192.350059  192.311726  192.318540  192.420677  198.241917  198.250296   
639    95.161885   95.142921   95.146292   95.196823   98.076781   98.080927   
730   188.260303  188.222785  188.229454  188.329419  194.026888  194.035089   
424   963.908336  963.716243  963.750388  964.262220  993.433729  993.475722   
700    75.860269   75.845152   75.847839   75.888120   78.183939   78.187244   
517   254.820683  254.769901  254.778927  254.914236  262.626073  262.637174   
331   196.991317  196.952060  196.959038  197.063640  203.025342  203.033924   
1442  228.355319  228.309811  228.317900  228.439156  235.350051  235.359999   
1624  156.164181  156.133060  156.138592  156.221514  160.947633  160.954436   
877   163.105295  163.072791  163.078569  163.165177  168.101360  168.108465   
1880  255.939409  255.888404  255.897470  256.033373  263.779067  263.790217   
1238  162.085319  162.053018  162.058760  162.144827  167.050141  167.057202   
1277   68.122685   68.109109   68.111522   68.147695   70.209345   70.212313   
532   107.505026  107.483602  107.487410  107.544495  110.798003  110.802687   
1179   56.328416   56.317191   56.319186   56.349097   58.053807   58.056261   
1110  113.959803  113.937092  113.941129  114.001641  117.450496  117.455460   
...          ...         ...         ...         ...         ...         ...   
2595  124.528671  124.503855  124.508266  124.574390  128.343098  128.348524   
245    74.737371   74.722477   74.725124   74.764809   77.026645   77.029901   
2006  209.697547  209.655757  209.663185  209.774534  216.120774  216.129910   
2220  308.054328  307.992937  308.003849  308.167425  317.490313  317.503734   
2108  226.665090  226.619919  226.627948  226.748306  233.608049  233.617923   
439    69.982093   69.968147   69.970626   70.007786   72.125709   72.128757   
285   194.862315  194.823482  194.830385  194.933856  200.831126  200.839616   
2011  124.541421  124.516602  124.521013  124.587144  128.356239  128.361664   
896    98.218541   98.198967   98.202446   98.254600  101.227064  101.231343   
2500   69.688370   69.674483   69.676951   69.713955   71.822989   71.826025   
495   300.878630  300.818670  300.829328  300.989093  310.094818  310.107926   
1465   89.222651   89.204870   89.208030   89.255407   91.955622   91.959509   
1858  121.313203  121.289027  121.293324  121.357741  125.029137  125.034422   
545   150.195111  150.165180  150.170500  150.250253  154.795725  154.802268   
489    84.892775   84.875857   84.878865   84.923942   87.493119   87.496817   
690   186.237814  186.200700  186.207297  186.306188  191.942449  191.950562   
339   168.381440  168.347884  168.353849  168.443258  173.539117  173.546453   
1900   97.228839   97.209463   97.212907   97.264535  100.207048  100.211283   
856   265.823

In [175]:
#sample = results_multiplier.sample(frac=0.2,axis=0)
model_err = metrics.median_absolute_error(sample.iloc[:,:-1].values.flatten(), new_predictions.iloc[:,:-1].values.flatten())

In [176]:
model_err

72.1070818480257

## Using ARIMA time series models for future forecasting